File: platforms/nodejs/src/socketClient.js
'use strict';

const {io} = require('socket.io-client');

const NetMsg = {
    Connect: 'connect',
    Disconnect: 'disconnect',
    Handshake: 'handshake',
    Error: 'error',
    NewBlockHeaders: 'newBlockHeaders',
    Logs: 'logs',
    FilteredTrades: 'filteredTrades',
    FilteredPairs: 'filteredPairs'
};

class ShardsSocketClient {
    constructor(_localizedMatcher=null) {
        this._localizedMatcher = _localizedMatcher;
        this._url = 'wss://wiretap.web3shards.io';
        this._hook = this._hook.bind(this);
        this._connect = this._connect.bind(this);

    }

    /**
     * Creates a new websocket connection
     * @param {*} _config Config: {chain,shardsApiKey,shardsWebsocketId}
     * @param {*} _localizedMatcher Object with fields to filter incoming data by.
     * @returns 
     */
    createConnection(_config, _localizedMatcher=null) {
        let _conn = new ShardsSocketClient(_localizedMatcher);
        return _conn._connect(_config);
    }

    subscribe(_hook) {
        this._emitData = _hook;
        return this;
    }

    onDisconnect(_hook) {
        this._emitDisconnect = _hook;
        return this;
    }

    onError(_hook) {
        this._emitError = _hook;
        return this;
    }

    _connect(_config) {
        try {
            this._config = _config;
            console.log('connecting to web3 shards @ '+this._url);
            this._socket = io(this._url, {
                reconnectionDelayMax: 10000,
                path: `/${this._config.chain}/`
            });
            this._socket.on(NetMsg.Connect, this._hook);
        } catch (_err) {
            console.log(`an error occurred: ${_err}`);
        }
        return this;
    }

    _hook() {
        this._socket.removeAllListeners();
        this._socket.on(NetMsg.Connect, this._hook);
        
        this._socket.on(NetMsg.Handshake, function(_data) {
            if (_data.success) {
                console.log(`handshake confirmed. waiting for data...`);
            }
        }.bind(this));

        this._socket.on(NetMsg.Error, function(_error) {
            if (this._emitError)
                this._emitError(`an error occurred: ${_error.error}`);
        }.bind(this));

        this._socket.on(NetMsg.Disconnect, function() {
            if (this._emitDisconnect)
                this._emitDisconnect();
        }.bind(this));
        
        this._socket.on(NetMsg.NewBlockHeaders, function(_data) {
            if (this._emitData)
                this._emitData(NetMsg.NewBlockHeaders, _data);
        }.bind(this));
        
        this._socket.on(NetMsg.Logs, function(_data) {
            if (this._emitData)
                this._emitData(NetMsg.Logs, _data);
        }.bind(this));
        
        this._socket.on(NetMsg.FilteredTrades, function(_data) {
            if (this._emitData)
                this._emitData(NetMsg.FilteredTrades, _data);
        }.bind(this));
        
        this._socket.on(NetMsg.FilteredPairs, function(_data) {
            if (this._emitData)
                this._emitData(NetMsg.FilteredPairs, _data);
        }.bind(this));

        let _handshake = {
            apiKey: this._config.shardsApiKey,
            roomId: this._config.shardsWebsocketId
        }

        if (this._localizedMatcher) {
            _handshake.match = this._localizedMatcher;
        }

        this._socket.emit(NetMsg.Handshake, _handshake);
    }

}

module.exports = new ShardsSocketClient();

File: platforms/python/3.12.2/socket_client.py
import asyncio
import socketio
import time
from netmsg import NetMsg
from chains import Chain

# Create a receiver function
def data_handler(dataType, data):
    print("received data:", dataType, ": ", data)

# Initialize socketio and shards settings
sio = socketio.Client(logger=False)
shardsSocketSettings={
    "url": "wss://wiretap.web3shards.io",
    "chain": Chain.Ethereum.value,
    "auth": {
        "shardsApiKey": "",
        "shardsWebsocketId": ""
    },
    "subscriber": data_handler
}

# Setup event handlers
@sio.on(NetMsg.Connect.value)
def _on_connect():
    print("[ShardsSocketClient]", NetMsg.Connect.value)
    sio.emit(NetMsg.Handshake.value, {
        "apiKey": shardsSocketSettings["auth"]["shardsApiKey"],
        "roomId": shardsSocketSettings["auth"]["shardsWebsocketId"]
    })

@sio.on(NetMsg.Disconnect.value)
def _on_disconnect():
    print("[ShardsSocketClient]", NetMsg.Disconnect.value)

@sio.on(NetMsg.Error.value)
def _on_error(err):
    print("[ShardsSocketClient]", NetMsg.Error.value, ":", err)

@sio.on(NetMsg.Handshake.value)
def _on_handshake(data):
    print("[ShardsSocketClient] authenticated. waiting for data...")

@sio.on(NetMsg.FilteredTrades.value)
def _on_filtered_trades(data):
    shardsSocketSettings["subscriber"](NetMsg.FilteredTrades.value, data)

# Create listener
def listen():
    url = shardsSocketSettings["url"]
    path = "/"+shardsSocketSettings["chain"]
    sio.connect(url, socketio_path=path, transports=["websocket"])
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        print("execution interrupted. disconnecting")
        sio.disconnect()
        
listen()